resteasy-core/src/main/java/org/jboss/resteasy/core/ExceptionHandler.java

```diff
@@ -165,7 +165,16 @@ protected Response unwrapException(HttpRequest request, Throwable e, RESTEasyTra
 
       if (unwrappedException instanceof WebApplicationException) {
          WebApplicationException wae = (WebApplicationException) unwrappedException;
-         if (wae.getResponse() != null && wae.getResponse().getEntity() != null) return wae.getResponse();
+         Response response = wae.getResponse();
+         if (response != null) {
+            try {
+               if (response.getEntity() != null) return response;
+            }
+            catch(IllegalStateException ise) {
+               // IllegalStateException from ClientResponse.getEntity() means the response is closed and got no entity
+               return response;

```

Comment:
@lyind, it seems to me that the flow of control should drop through here, rather than returning the response. The spec (3.3.4 Exceptions, item 1) says
```
If the response property of the exception does not contain an entity and an exception mapping
provider (see Section 4.4) is available for WebApplicationException or the corresponding sub-
class, an implementation MUST use the provider to create a new Response instance, otherwise the
response property is used directly.
```

If the Response is closed, then, in effect, it doesn't have an entity.  Falling through gives us a chance to check if "an exception mapping provider (see Section 4.4) is available for WebApplicationException or the corresponding sub-class".

-Ron

Reply:
Right, corrected in bd16649e45b4177a61a6f6c4d882b8002b66b584.